keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,292 @@
+a new feature-flag named `kustomize`, initially alpha and disabled by default.
+
+#### Kustomize integration with kubeadm
+
+By adopting Kustomize in the context of Kubeadm, this proposal assumes to:
+
+- Let kubeadm generate static pod manifests or kubelet configuration as usual.
+- Use kubeadm generated artifacts as a starting point for applying patches
+  containing “advanced configurations”.
+
+This has some implications:
+
+1. The higher-level tools/users have to express “advanced configurations” using
+   one of the two alternative techniques supported by Kustomize - the [strategic
+   merge patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchstrategicmerge)
+   and the [JSON patch](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/glossary.md#patchjson6902) -.
+2. The higher-level tools/users have to provide patches before running kubeadm;
+   this point is going to be further discussed in the following paragraphs.
+3. Kubeadm is responsible for coordinating the execution of Kustomize within the
+   init/join/upgrade workflows
+4. as a consequence of the previous point, higher-level tools/users are not
+   requested to take care of defining `kustomization.yaml` files nor to define
+   a local folder structure.
+
+Additionally, in order to simplify the first implementation of this KEP, this 
+proposal is going to assume that Kustomize patches for kubeadm are always defined
+specifically for the node where kubeadm is going to be executed.
+
+This point could be reconsidered in the future, by e.g. introducing cluster-wide
+patches and/or patches for a subset of nodes.
+
+#### Providing and storing Kustomize patches to kubeadm
+
+Before kubeadm init, Kustomize patches should be eventually provided to kubeadm
+by higher-level tools/users; patches should be provided in a well know location
+like e.g. `/etc/kubernetes/kustomize` or in a custom location specified with a CLI flag. 
+Eventually, in the future, this flag will be added to the kubeadm config API as well.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+to use the same approach also for kubeadm join; this point could be reconsidered
+in the future, by e.g. defining a method for allowing higher-level tools/users to
+define Kustomize patches using a new CRD.
+
+#### Storing and retrieving Kustomize patches for kubeadm
+
+Kustomize patches, should be preserved during the whole cluster lifecycle, mainly
+for allowing kubeadm to preserve changes during the kubeadm upgrade workflow.
+
+In order to simplify the first implementation of this KEP, this proposal is assuming
+that Kustomize patches will remain stored in the `/etc/kubernetes/kustomize` or in a
+custom location specified with a CLI flag for the necessary time; this point could
+be reconsidered in the future, by e.g. defining a method for allowing higher-level
+tools/users to define Kustomize patches using a new CRD.
+
+### Risks and Mitigations
```

Comment:
i think we need to cover the case - what happens if a patch no longer applies after upgrade.

Reply:
done